### PR TITLE
[WIP] Almost complete the shortcut

### PR DESCRIPTION
**Pull request by yiliey**
I have improved the matrix implementation for IF,rps and tracin, where the changes can be found in the main function and testing code will also be found in the testing files. Though for trak, I only simplify the calculation of testing gradients and I find it hard to further simplify.

### CODE REVIEW

File: dattri/algorithm/tracin.py [hunk 1]
```diff
@@ -215,3 +216,136 @@ def attribute(
...
+                    # accumulate the TDA score in corresponding positions (blocks)
+                if self.normalized_grad:
+                    tda_output[row_st:row_ed] += (
+                        (
+                            torch.ones(row_ed - row_st)
+                            * ckpt_weight
+                        )
+                        .detach()
+                        .cpu()
+                    )
```
[the remark this answers is not in the document]
Comment (yiliey) [reply dattri/algorithm/tracin.py:343]: Maybe I need row_ed - row_st here so I can't move it to the front.
